### PR TITLE
refactor(core): remove ext: modules from the module map

### LINKED ISSUE #17717: prevent importmap to acess internal specifiers

### PR DESCRIPTION
**Pull request by nayeemrmn**
Rather than disallowing `ext:` resolution, clear the module map after initializing extensions so extension modules are anonymized. This operation is explicitly called in `deno_runtime`. Re-inject `node:` specifiers into the module map after doing this.

Fixes #17717.

### CODE REVIEW

File: cli/resolver.rs
Comment (bartlomieju) [inline cli/resolver.rs:38]: Are we 100% sure there are no scenarios where rejecting on load is too late?

Also, any ideas how you'd handle `node:` specifiers? Maybe we should add an "alias" system for the `ext:` specifiers?

File: cli/resolver.rs
Comment (bartlomieju) [inline cli/resolver.rs:242]: Fixes https://github.com/denoland/deno/issues/17717?

File: cli/tests/testdata/run/extension_import.ts.out
Comment (bartlomieju) [inline cli/tests/testdata/run/extension_import.ts.out:2]: This is much better error to be honest.

File: core/modules.rs
Comment (bartlomieju) [inline core/modules.rs:413]: I don't understand this bit - why would we have `ModuleMap` created before `ExtModuleLoader`?

File: core/modules.rs
Comment (bartlomieju) [inline core/modules.rs:444]: Could you improve this message in `expect`? I can't really understand its meaning

File: core/runtime.rs
Comment (bartlomieju) [inline core/runtime.rs:509]: Why do we need to store this callback here? Is it in preparation for modules support in realms?

File: core/modules.rs
Comment (nayeemrmn) [reply core/modules.rs:413]: It's so `ExtModuleLoader` is made aware of extension modules that are already in the snapshot, since those aren't passed in `Extension::get_esm_sources()`. Maybe that info shouldn't be excluded... we can add `ExtensionFileSourceCode::IncludedInSnapshot` which asserts that it never gets loaded.

File: core/runtime.rs
Comment (nayeemrmn) [reply core/runtime.rs:509]: I guess it's needed to support maybe spawning a realm in the bootstrap code. We'll very likely never do that, but it's also simpler to have this field than to pass it as an argument to `init_extension_js()` so it can be absent on realms.

File: core/modules.rs
Comment (bartlomieju) [reply core/modules.rs:413]: I'd be +1 on that - the more side wheel to ensure nothing goes wrong the better

File: cli/resolver.rs
Comment (bartlomieju) [reply cli/resolver.rs:38]: Fair enough, any thoughts on `node:` specifiers?

Comment (nayeemrmn) [reply cli/resolver.rs:38]: I figure we can store a list of aliases on the extension. My first way was to store that information on each extension module source / module map entry, but this way doesn't have to preserve it through snapshot layers since extension metadata is passed at every step.

File: runtime/lib.rs
Comment (nayeemrmn) [inline runtime/lib.rs:41]: `create_cli_snapshot()` didn't list `runtime` as an extension and had been implicitly depending on it from the built-upon snapshot, that needed to be added since we access to the extension's metadata across snapshots.

Turns out the `runtime` extensions was defined in `build.rs` and cli couldn't import it, I had to duplicate it.

File: core/extensions.rs
Comment (nayeemrmn) [inline core/extensions.rs:30]: `ExtensionFileSourceCode::AlreadyEvaluated`. Previously this was represented by just not telling core about esm files that are already evaluated in the passed snapshot, but we need access to the total list of ext specifier names so now it's represented by this variant.

I think supporting `ShadowRealm` will require all extension code at runtime though, in addition to the startup snapshot, which would incur some binary cost.

File: cli/resolver.rs
Comment (nayeemrmn) [reply cli/resolver.rs:38]: I thought of another approach. All of the ideas we've discussed so far require `deno_core` to guess at what the user wants to do with the extension module map based on whether or not a snapshot will be created, and are cryptically specific to CLI's usage. Instead I think we should have explicit procedural bindings for those behaviours. They will be very niche but it's unavoidable because that's the granularity we need:
- Keep all extension modules in the module map by default.
- Add the following APIs to `JsRuntime`:
  ```rust
  impl JsRuntime {
    /// Clear the internal module map.
    fn clear_module_map(&self) {}
    /// Get a module handle by specifier from the internal module map.
    fn get_module_handle(&self, specifier: &Url) -> Option<v8::Global<v8::Module>> {}
    /// Inject a module handle by specifier into the internal module map. The module must be
    /// kEvaluated. It will be stored with no dependency references.
    fn inject_module_handle(&self, specifier: Url, v8::Global<v8::Module>) {}
  }
  ```
- Create a runtime worker roughly like this:
  ```rust
  fn get_ext_node_module_handles(js_runtime: &JsRuntime) -> Vec<(Url, v8::Global<v8::Module>)> {
    vec![
      ("node:assert", js_runtime.get_module_handle("ext:deno_node/assert.ts").unwrap()),
      ("node:buffer", js_runtime.get_module_handle("ext:deno_node/buffer.ts").unwrap()),
      // ...
    ]
  }
  let js_runtime = JsRuntime::new(options);
  let ext_node_module_handles = get_ext_node_module_handles(&js_runtime);
  js_runtime.clear_Module_map();
  for (specifier, handle) in ext_node_module_handles {
    js_runtime.inject_module_handle(specifier, handle);
  }
  ```

File: runtime/worker.rs
Comment (nayeemrmn) [inline runtime/worker.rs:76]: @bartlomieju This is how we make extension modules non-importable except for aliasing and exposing `ext:deno_node/` as `node:` specifiers. It's done in `deno_runtime`.

When realm-specific module maps are supported (#15760), we would do this for every realm with a callback and it would look cleaner overall:
```rust
JsRuntime::new(RuntimeOptions {
  // ...
  realm_init_cb: |js_realm| {
    let handles = deno_node::SUPPORTED_BUILTIN_NODE_MODULES
      .iter()
      .map(|p| (p, js_realm.get_module_handle(p.ext_specifier).unwrap()))
      .collect::<Vec<_>>();
    js_realm.clear_module_map();
    for (p, handle) in handles {
      js_realm.inject_module_handle(
        deno_core::FastString::from_static(p.specifier),
        deno_core::ModuleType::JavaScript,
        handle,
      )
    }
  }
})
```

File: core/extensions.rs
Comment (bartlomieju) [reply core/extensions.rs:30]: Why do you think we need to require all extension code for `ShadowRealm`? Wouldn't we start up from the same snapshot there? 

Snapshots can store multiple different contexts with different globals variables - we can leverage this feature to selectively make some extensions available.

File: core/extensions.rs
Comment (bartlomieju) [inline core/extensions.rs:373]: What does this TODO mean?

File: ext/node/polyfill.rs
Comment (bartlomieju) [inline ext/node/polyfill.rs:6]: So this implies that `module_name` passed to this function doesn't have `node:` prefix - Could you add a docstring about it to this function?

File: runtime/lib.rs
Comment (bartlomieju) [reply runtime/lib.rs:41]: Could you explain the reasoning here a bit? Is it done just for the sake of ensuring extension ordering?

File: runtime/worker.rs
Comment (bartlomieju) [reply runtime/worker.rs:76]: This is quite an elegant solution, but I don't like that it's in `runtime/worker.rs`. How about we make it an API in `JsRuntime` that takes a `Vec<(String, String)>` of `(specifier, ext_specifier)` and does this mapping internally.

JsRuntime::inject_module_handle` seems a bit too low-level to expose it (for the time being at least)

**Review by bartlomieju — COMMENTED** [review]
Sorry for a slow response @nayeemrmn, I've been busy preparing for v1.34 release. I really like the direction of this PR and I believe it cleans up a lot of concepts.

That said I need to take another look at `core/modules.rs` with fresh eyes before approving and landing it.

I also left some comments with questions and suggestions; but overall it's great.

File: ext/node/polyfill.rs
Comment (dsherret) [inline ext/node/polyfill.rs:6]: Perhaps we could add a helper function to `NodeModulePolyfill` that does this?

```rs
impl NodeModulePolyfill {
  pub fn module_name(&self) -> &'static str {
    debug_assert!(self.name.starts_with("node:"));
    &self.name[5..]
  }
}
```

File: core/extensions.rs
Comment (nayeemrmn) [reply core/extensions.rs:30]: > Snapshots can store multiple different contexts with different globals variables

My understanding of this is that you can have bootstrapping code like `globalThis.friendRealm = new ShadowRealm()`, snapshot that, load the snapshot and still have both realms. But there's no way of using an isolate-wide startup snapshot to start a new realm in that same isolate, is there?

File: runtime/lib.rs
Comment (nayeemrmn) [reply runtime/lib.rs:41]: This and the `ExtensionFileSourceCode::AlreadyEvaluated` stuff is so `ExtModuleLoader::resolve()` has access to the full list of extension module specifiers so it knows what to `Ok()` as is and what to defer to the inner module loader. I think I'm going to revert that part since it isn't necessary with the current approach. Actually we have no use for giving `ExtModuleLoader` access to the user's module loader as a fallback... we can just remove that and `Ok()` everything in `ExtModuleLoader::resolve()`.

File: runtime/worker.rs
Comment (nayeemrmn) [reply runtime/worker.rs:76]: > How about we make it an API in `JsRuntime` that takes a `Vec<(String, String)>` of `(specifier, ext_specifier)` and does this mapping internally.

The behaviour we want is so specific I can't begin to imagine what this API would be called. IMO, that API and others we've considered are obfuscations of this explicit approach which expose more-or-less the same capability.

File: core/extensions.rs
Comment (andreubotella) [reply core/extensions.rs]: Snapshots seem to be of a context, not of an isolate as a whole somehow; and in my testing, it seemed like if there is a snapshot for the isolate, that will be applied by default when a context is created.

After all, V8 uses snapshots internally for initializing the JS built-ins (though written in CSA and Torque, not JS), and creating a context creates different built-ins – in terms of object identity, at least.

Comment (bartlomieju) [reply core/extensions.rs]: Andreu is right here - currently we use the first approach (ie. the snapshot of a isolate), but we have all the necessary APIs to have snapshots of different contexts. @crowlKats and I planned to experiment with the second approach so I think we shouldn't discard it yet. But definitely something that can be addressed in a follow up.

File: runtime/worker.rs
Comment (bartlomieju) [reply runtime/worker.rs:76]: `JsRuntime::alias_known_modules` would work. Sure, it's specific, but I think it's better than exposing so low level API that allows you to manipulate module handles.

File: core/extensions.rs
Comment (nayeemrmn) [reply core/extensions.rs]: Good to know, I think that means my `This should be per-realm` TODOs aren't applicable

File: runtime/worker.rs
Comment (mmastrac) [inline runtime/worker.rs:65]: I think this rename and retain operation callback should live in the extension definition macro (deno_core::extension!) itself.

File: ext/node/polyfill.rs
Comment (mmastrac) [inline ext/node/polyfill.rs:19]: strip_prefix and unwrap

File: core/runtime.rs
Comment (mmastrac) [inline core/runtime.rs:1845]: See below - this should be called internally based on extension macro config

File: core/runtime.rs
Comment (mmastrac) [inline core/runtime.rs:680]: Yeah, this code is unfortunately quite broken with realms.

**Review by mmastrac — CHANGES_REQUESTED** [review]
I love these changes and love to see so many special cases deleted.

 I think there's only one nit and one change I'd like to see -- moving this new retain/rename operation so that it's attached to the extension macro itself and called after JS init, which helps decouple the code.

File: ext/node/polyfill.rs
Comment (dsherret) [reply ext/node/polyfill.rs:19]: It was previously that https://github.com/denoland/deno/pull/19040#discussion_r1203158425

I think either way is fine because all of these should start with "node:"

File: runtime/worker.rs
Comment (nayeemrmn) [reply runtime/worker.rs:65]: > I think this rename and retain operation callback should live in the extension definition macro (deno_core::extension!) itself.

That was my first idea. The problem is we don't want this step to be taken when we're creating the intermediary `RUNTIME_SNAPSHOT` since the CLI extensions will want internal access. While thinking of how on earth I'd document or option-ise this very specific behaviour, it became clear to me that it's not a property of the extension but rather a bespoke operation called by the host only at the final step. So I added a procedural binding for it.

IMO It turns out that this solution is the less coupled one and is easier to delete if e.g. we figure out a better cross-crate JS extension interface.

Comment (mmastrac) [reply runtime/worker.rs:65]: That makes sense to me. I think we can take another look at the runtime/realm setup process after this lands (especially if we can get modules working correctly in realms soon).

**Comment by mmastrac** [comment]
After reviewing the code and feedback I'm good with this as-is. I'll approve this as-is. Great PR.

**Review by mmastrac — APPROVED** [review]

**Pull request merged**
